### ISSUE TITLE
Remote-SSH does not use ssh config defaults, corrupts config file

### ISSUE THREAD

**Issue by 4wrxb** [microsoft/vscode-remote-release#4592]
Issue Type: <b>Bug</b>

The Remote-SSH extesion is having issues with the SSH config file. I am on Windows using a standard git-bash install. The git config in %USERPROFILE%\.ssh\config is a well-formed file which has host aliases followed by domain-based defaults followed by user defaults. This has been confirmed working with the ssh and scp commands from git-bash. Remote-SSH respects none of these options and requires that all are within the alias host section (i.e. with a HostName option).

Further, adding a host through the VS Code GUI ignores the ordering of the .ssh/config file and appends the host section. OpenSSH order specifies that specific sections should be first and default sections should be last.

See [vscode_example_ssh.config.txt](https://github.com/microsoft/vscode-remote-release/files/6078421/vscode_example_ssh.config.txt) for an example. Other than replacing usernames and hostnames/domains it matches what I am currently using.

Extension version: 0.64.0
VS Code version: Code 1.53.2 (622cb03f7e070a9670c94bae1a45d78d7181fbd4, 2021-02-11T11:48:04.245Z)
OS version: Windows_NT x64 10.0.18363
Remote OS version: Linux x64 4.4.121-92.104-default

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz (4 x 2712)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.84GB (5.76GB free)|
|Process Argv|--crash-reporter-id 2a3683de-98a3-4030-b204-5ad30de37137|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: redacted|
|OS|Linux x64 4.4.121-92.104-default|
|CPUs|Intel Core Processor (Broadwell, IBRS) (2 x 3792)|
|Memory (System)|3.69GB (1.47GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstry914:30256638
pythonvsdeb440:30248342
pythonvsded773:30248341
pythonvspyt875:30259475
dockersubset:30265998
pythontb:30265425
vspre833cf:30267465

```

</details>

**Comment by roblourens** [comment]
> Remote-SSH respects none of these options

When connecting to a remote, the extension doesn't parse your config file. It simply runs something like `ssh <yourhost>` and ssh is parsing your config file.

> OpenSSH order specifies that specific sections should be first and default sections should be last.

Where does it say that?

**Comment by roblourens** [comment]
Hmm... I see. I guess I should add the new entry to be before the first wildcard entry.

**Comment by 4wrxb** [comment]
I would suggest simply putting it before existing all host or match keywords. Having dealt with scripting ssh configs once before it's tricky to detect all possible wildcard patterns (let alone multi-pattern sections etc.) and once CanonicalizeHostname or other options get thrown in the mix things can get even weirder.

Since the entry added by remote-ssh is known to be a simple single-pattern non-wildcard host entry it will be safe as the first _host/match entry_. (It's bad form, but there are examples of non-qualified configurations at the beginning of config files. Adding a host entry before these will change the behavior.)

**Comment by roblourens** [comment]
That would be easy. Especially for someone with a simple ssh config file, that's probably not what they expect though.

**Comment by roblourens** [comment]
I think it shouldn't go exactly at the top - it should go before the first Host/Match directive, and after any Include directives at the top of the file. See https://github.com/microsoft/vscode-remote-release/issues/4339#issuecomment-789418848 also

**Comment by roblourens** [comment]
I agree with your examples. I don't think we should do a setting yet. It seems like the ordering is more complicated than I originally realized, and your example is the only "safe" way to do it.
